drivers/src/mailbox.rs

```diff
@@ -22,57 +15,93 @@ caliptra_err_def! {
     Mailbox,
     MailboxErr
     {
-        // Invalid state
-        InvalidStateErr = 0x1,
         // Exceeds mailbox capacity
-        InvalidDlenErr = 0x2,
-        // No data avaiable.
-        NoDataAvailErr = 0x03,
-        // Enqueue Error
-        EnqueueErr = 0x04,
-        // Dequeue Error
-        DequeueErr = 0x05,
+        InvalidDlenErr = 0x1,
+        EnqueueErr = 0x02,
+        MailboxAccessErr = 0x03,
     }
 }
 
-#[derive(Copy, Clone, Default, Eq, PartialEq)]
-/// Malbox operational states
-pub enum MailboxOpState {
-    #[default]
-    RdyForCmd,
-    RdyForDlen,
-    RdyForData,
-    Execute,
-    Idle,
-}
-
-#[derive(Default, Debug)]
+/// Mailbox operational states.
+/// These are used to implement the typestate pattern.
+/// See https://rust-embedded.github.io/book/patterns/typestate.html
+/// for more information.
+/// The typestate pattern is used to statically enforce the correct
+/// order of mailbox operations.
+/// For example, the MailboxSendTxn type can only transition from
+/// the RdyForCmd state to the RdyForDlen state.
+/// This prevents the user from writing a data length before a command.
+///
+/// The typestate pattern is implemented using the following structs.
+/// Each struct represents a state and contains a method for each
+/// possible transition to another state.
+/// The method names are the same as the next state.
+/// For example, the RdyForCmd struct contains a method called
+/// write_cmd that transitions to the RdyForDlen state.
+
+/// Idle is the initial state of the mailbox, when it is unlocked.
+/// The mailbox is locked when a transaction is in progress.
+/// The mailbox is unlocked when the transaction is complete.
+///
+/// Idle is the initial state of the mailbox, when it is unlocked.
+pub struct Idle;
+/// RdyForCmd is the state after the mailbox is idle.
+pub struct RdyForCmd;
+pub struct RdyForDlen;
+pub struct RdyForData;
+pub struct Execute;
+#[derive(Default)]
 /// Caliptra mailbox abstraction
 pub struct Mailbox {}
 
 const MAX_MAILBOX_LEN: u32 = 128 * 1024;
 
 impl Mailbox {
+    pub fn send_request(&self, cmd: u32, data: &[u8]) -> CaliptraResult<MailboxSendTxn<Execute>> {
```

Comment:
I'm likely missing context, but why can't the send functionality be a single method?

```rust
pub fn send_request(&mut self, cmd: u32, data: &[u8]) -> CaliptraResult<()> {
    // Do sanity checks, e.g., is locked, is `data.len()` <= 128 KiB.

    // Block until mailbox is ready for a command.
    
    // Write out (cmd, dlen, data).
    
    // "Send" by setting `execute` flag.

    // Block until mailbox is done sending.

    // "Done send" by clearing execute flag.

    Ok(())
}
```
...adjusted accordingly depending on what the blocking behavior should be to the caller.

Reply:
Hey Royce, 

Overall you want to overlap your I/O operation with something else.

The user can still block waiting for mailbox access by checking the return.
Checkout the usage model for  'send_request'  in  idev_id.rs and sha384acc_tests.rs.